File: src/shared/companion-crypto.ts
export interface CompanionEncryptedEnvelope {
  algorithm: 'A256GCM'
  keyId: string
  nonce: string
  ciphertext: string
}

export function companionEventAssociatedData(event: {
  eventId: string
  protocolVersion: number
  type: string
  entityType: string
  entityId: string
  revision: number
  occurredAt: string
}): string {
  return [
    'project-agent:event', event.eventId, event.protocolVersion, event.type,
    event.entityType, event.entityId, event.revision, event.occurredAt
  ].join(':')
}

export function companionCommandAssociatedData(command: {
  commandId: string
  protocolVersion: number
  type: string
  createdAt: string
}): string {
  return ['project-agent:command', command.commandId, command.protocolVersion, command.type, command.createdAt].join(':')
}

export function companionAttachmentAssociatedData(accountId: string, attachmentId: string): string {
  return ['project-agent:attachment', accountId, attachmentId].join(':')
}

function base64UrlEncode(bytes: Uint8Array): string {
  let binary = ''
  for (const byte of bytes) binary += String.fromCharCode(byte)
  return btoa(binary).replaceAll('+', '-').replaceAll('/', '_').replaceAll('=', '')
}

function base64UrlDecode(value: string): Uint8Array {
  const normalized = value.replaceAll('-', '+').replaceAll('_', '/')
  const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, '=')
  const binary = atob(padded)
  return Uint8Array.from(binary, (character) => character.charCodeAt(0))
}

function arrayBuffer(bytes: Uint8Array): ArrayBuffer {
  const copy = new Uint8Array(bytes.byteLength)
  copy.set(bytes)
  return copy.buffer
}

async function importAccountKey(encoded: string, usage: KeyUsage): Promise<CryptoKey> {
  const bytes = base64UrlDecode(encoded)
  if (bytes.byteLength !== 32) throw new Error('Companion E2EE key must be 256 bits.')
  return crypto.subtle.importKey('raw', arrayBuffer(bytes), { name: 'AES-GCM' }, false, [usage])
}

export function generateCompanionAccountKey(): string {
  const bytes = new Uint8Array(32)
  crypto.getRandomValues(bytes)
  return base64UrlEncode(bytes)
}

export async function companionAccountKeyId(encoded: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', arrayBuffer(base64UrlDecode(encoded)))
  return base64UrlEncode(new Uint8Array(digest)).slice(0, 16)
}

export async function sealCompanionBytes(
  encodedKey: string,
  plaintext: Uint8Array,
  associatedData: string
): Promise<CompanionEncryptedEnvelope> {
  const nonce = new Uint8Array(12)
  crypto.getRandomValues(nonce)
  const ciphertext = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: arrayBuffer(nonce), additionalData: arrayBuffer(new TextEncoder().encode(associatedData)), tagLength: 128 },
    await importAccountKey(encodedKey, 'encrypt'),
    arrayBuffer(plaintext)
  )
  return {
    algorithm: 'A256GCM',
    keyId: await companionAccountKeyId(encodedKey),
    nonce: base64UrlEncode(nonce),
    ciphertext: base64UrlEncode(new Uint8Array(ciphertext))
  }
}

export async function openCompanionBytes(
  encodedKey: string,
  envelope: CompanionEncryptedEnvelope,
  associatedData: string
): Promise<Uint8Array> {
  if (envelope.algorithm !== 'A256GCM') throw new Error('Unsupported Companion encryption algorithm.')
  if (envelope.keyId !== await companionAccountKeyId(encodedKey)) throw new Error('Companion encryption key does not match.')
  const nonce = base64UrlDecode(envelope.nonce)
  if (nonce.byteLength !== 12) throw new Error('Companion encryption nonce is invalid.')
  const plaintext = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: arrayBuffer(nonce), additionalData: arrayBuffer(new TextEncoder().encode(associatedData)), tagLength: 128 },
    await importAccountKey(encodedKey, 'decrypt'),
    arrayBuffer(base64UrlDecode(envelope.ciphertext))
  )
  return new Uint8Array(plaintext)
}

export async function sealCompanionJson(
  encodedKey: string,
  value: unknown,
  associatedData: string
): Promise<CompanionEncryptedEnvelope> {
  return sealCompanionBytes(encodedKey, new TextEncoder().encode(JSON.stringify(value)), associatedData)
}

export async function openCompanionJson<T>(
  encodedKey: string,
  envelope: CompanionEncryptedEnvelope,
  associatedData: string
): Promise<T> {
  return JSON.parse(new TextDecoder().decode(
    await openCompanionBytes(encodedKey, envelope, associatedData)
  )) as T
}

const attachmentMagic = new Uint8Array([0x50, 0x41, 0x45, 0x32]) // PAE2

export async function sealCompanionAttachment(
  encodedKey: string,
  plaintext: Uint8Array,
  associatedData: string
): Promise<Uint8Array> {
  const nonce = new Uint8Array(12)
  crypto.getRandomValues(nonce)
  const ciphertext = new Uint8Array(await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: arrayBuffer(nonce), additionalData: arrayBuffer(new TextEncoder().encode(associatedData)), tagLength: 128 },
    await importAccountKey(encodedKey, 'encrypt'),
    arrayBuffer(plaintext)
  ))
  const sealed = new Uint8Array(attachmentMagic.byteLength + nonce.byteLength + ciphertext.byteLength)
  sealed.set(attachmentMagic, 0)
  sealed.set(nonce, attachmentMagic.byteLength)
  sealed.set(ciphertext, attachmentMagic.byteLength + nonce.byteLength)
  return sealed
}

export async function openCompanionAttachment(
  encodedKey: string,
  sealed: Uint8Array,
  associatedData: string
): Promise<Uint8Array> {
  const minimumLength = attachmentMagic.byteLength + 12 + 16
  if (sealed.byteLength < minimumLength || !attachmentMagic.every((byte, index) => sealed[index] === byte)) {
    throw new Error('Companion attachment envelope is invalid.')
  }
  const nonce = sealed.slice(attachmentMagic.byteLength, attachmentMagic.byteLength + 12)
  const ciphertext = sealed.slice(attachmentMagic.byteLength + 12)
  return new Uint8Array(await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: arrayBuffer(nonce), additionalData: arrayBuffer(new TextEncoder().encode(associatedData)), tagLength: 128 },
    await importAccountKey(encodedKey, 'decrypt'),
    arrayBuffer(ciphertext)
  ))
}

File: ios/ProjectAgentCompanion/CompanionCrypto.swift
import CryptoKit
import Foundation

struct CompanionEncryptedEnvelope: Codable, Equatable {
    let algorithm: String
    let keyId: String
    let nonce: String
    let ciphertext: String
}

enum CompanionCrypto {
    private static let attachmentMagic = Data([0x50, 0x41, 0x45, 0x32])

    static func eventAssociatedData(
        eventId: String,
        protocolVersion: Int,
        type: CompanionEventType,
        entityType: String,
        entityId: String,
        revision: Int64,
        occurredAt: String
    ) -> String {
        [
            "project-agent:event", eventId, String(protocolVersion), type.rawValue,
            entityType, entityId, String(revision), occurredAt
        ].joined(separator: ":")
    }

    static func commandAssociatedData(
        commandId: String,
        protocolVersion: Int,
        type: CompanionCommandType,
        createdAt: String
    ) -> String {
        ["project-agent:command", commandId, String(protocolVersion), type.rawValue, createdAt]
            .joined(separator: ":")
    }

    static func attachmentAssociatedData(accountId: String, attachmentId: String) -> String {
        ["project-agent:attachment", accountId, attachmentId].joined(separator: ":")
    }

    static func sealJSON<T: Encodable>(_ value: T, key encodedKey: String, associatedData: String) throws -> CompanionEncryptedEnvelope {
        let plaintext = try JSONEncoder().encode(value)
        let key = try symmetricKey(encodedKey)
        let sealed = try AES.GCM.seal(plaintext, using: key, authenticating: Data(associatedData.utf8))
        return CompanionEncryptedEnvelope(
            algorithm: "A256GCM",
            keyId: try keyIdentifier(encodedKey),
            nonce: Data(sealed.nonce).base64URLEncodedString(),
            ciphertext: (sealed.ciphertext + sealed.tag).base64URLEncodedString()
        )
    }

    static func openJSON<T: Decodable>(
        _ type: T.Type,
        envelope: CompanionEncryptedEnvelope,
        key encodedKey: String,
        associatedData: String
    ) throws -> T {
        try JSONDecoder().decode(type, from: open(envelope: envelope, key: encodedKey, associatedData: associatedData))
    }

    static func sealAttachment(_ plaintext: Data, key encodedKey: String, associatedData: String) throws -> Data {
        let sealed = try AES.GCM.seal(
            plaintext,
            using: symmetricKey(encodedKey),
            authenticating: Data(associatedData.utf8)
        )
        return attachmentMagic + Data(sealed.nonce) + sealed.ciphertext + sealed.tag
    }

    static func openAttachment(_ envelope: Data, key encodedKey: String, associatedData: String) throws -> Data {
        guard envelope.count >= attachmentMagic.count + 12 + 16,
              envelope.prefix(attachmentMagic.count) == attachmentMagic else {
            throw RelayError.integrity("附件加密信封无效。")
        }
        let body = envelope.dropFirst(attachmentMagic.count)
        let nonce = body.prefix(12)
        let ciphertextAndTag = body.dropFirst(12)
        guard ciphertextAndTag.count >= 16 else { throw RelayError.integrity("附件加密信封无效。") }
        let sealedBox = try AES.GCM.SealedBox(
            nonce: AES.GCM.Nonce(data: nonce),
            ciphertext: ciphertextAndTag.dropLast(16),
            tag: ciphertextAndTag.suffix(16)
        )
        return try AES.GCM.open(
            sealedBox,
            using: symmetricKey(encodedKey),
            authenticating: Data(associatedData.utf8)
        )
    }

    private static func open(envelope: CompanionEncryptedEnvelope, key encodedKey: String, associatedData: String) throws -> Data {
        guard envelope.algorithm == "A256GCM",
              envelope.keyId == (try keyIdentifier(encodedKey)),
              let nonce = Data(base64URLEncoded: envelope.nonce), nonce.count == 12,
              let ciphertextAndTag = Data(base64URLEncoded: envelope.ciphertext), ciphertextAndTag.count >= 16 else {
            throw RelayError.integrity("Companion 加密信封无效或密钥不匹配。")
        }
        let sealedBox = try AES.GCM.SealedBox(
            nonce: AES.GCM.Nonce(data: nonce),
            ciphertext: ciphertextAndTag.dropLast(16),
            tag: ciphertextAndTag.suffix(16)
        )
        return try AES.GCM.open(
            sealedBox,
            using: symmetricKey(encodedKey),
            authenticating: Data(associatedData.utf8)
        )
    }

    private static func symmetricKey(_ encoded: String) throws -> SymmetricKey {
        guard let bytes = Data(base64URLEncoded: encoded), bytes.count == 32 else {
            throw RelayError.integrity("同步密钥无效，请退出 Fuddy 后重新登录。")
        }
        return SymmetricKey(data: bytes)
    }

    private static func keyIdentifier(_ encoded: String) throws -> String {
        guard let bytes = Data(base64URLEncoded: encoded), bytes.count == 32 else {
            throw RelayError.integrity("同步密钥无效，请退出 Fuddy 后重新登录。")
        }
        return Data(SHA256.hash(data: bytes)).base64URLEncodedString().prefix(16).description
    }
}

private extension Data {
    init?(base64URLEncoded value: String) {
        var normalized = value.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        normalized += String(repeating: "=", count: (4 - normalized.count % 4) % 4)
        self.init(base64Encoded: normalized)
    }

    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
